src/deriveNativeScriptHash.c

```diff
@@ -0,0 +1,520 @@
+#define HEADER ui_native_script_header[ctx->ui_scriptType]
+
+#define SPRINTF_POSITION(offset, str) \
+	offset += snprintf(&str[offset], 10 + 1, "Position: ");\
+	for (int i = 1; i <= level; i++) {\
+		snprintf(&str[offset], 11 + 1, "%u.", ctx->complexScripts[i].totalScripts - ctx->complexScripts[i].remainingScripts + 1);\
+		/* snprintf returns 0, https://github.com/LedgerHQ/nanos-secure-sdk/issues/28 */\
+		offset += strlen(&str[offset]);\
+	}
+
+static void deriveScriptHash_display_ui_complexScriptFinished(ui_callback_fn_t* callback)
+{
+	ASSERT_UI_SCRIPT_TYPE_SANITY();
+	TRACE();
+
+	uint8_t level = ctx->level - 1;
+	if (level == 0) {
+		// Don't show position for complex script at level 0, "root" level
+		ui_displayPaginatedText(
+		        HEADER,
+		        "Finished",
+		        callback
+		);
+		return;
+	}
+
+	// "Position: x.x.x - Finished"
+	char content[10 + 11 * level + 10 + 1];
+	int offset = 0;
+	SPRINTF_POSITION(offset, content);
+
+	// replace trailing '.' with ' '
+	content[offset - 1] = ' ';
+
+	snprintf(&content[offset], 10 + 1, "- Finished");
+
+	ui_displayPaginatedText(
+	        HEADER,
+	        content,
+	        callback
+	);
+}
+
+static void deriveScriptHash_display_ui_position(uint8_t level, ui_callback_fn_t* callback)
+{
+	ASSERT_UI_SCRIPT_TYPE_SANITY();
+	TRACE();
+
+	// For level n the longest possible position is: "x.x.x...x"
+	// where the x is 2^32-1 (length 10) and there are up to level xs
+	// + the leading "Position: "
+	char position[10 + 11 * level + 1];
+	int offset = 0;
+	SPRINTF_POSITION(offset, position);
+
+	// remove any trailing '.'
+	position[offset - 1] = '\0';
+
+	ui_displayPaginatedText(
+	        HEADER,
+	        position,
+	        callback
+	);
+}
+
+#undef SPRINTF_POSITION
+
+enum {
+	DISPLAY_UI_STEP_COMPLEX_SCRIPT_FINISHED = 100,
+	DISPLAY_UI_STEP_POSITION = 200,
+	DISPLAY_UI_STEP_SCRIPT_CONTENT,
+	DISPLAY_UI_STEP_RESPOND,
+	DISPLAY_UI_STEP_INVALID
+};
+
+static void deriveScriptHash_display_ui_runStep()
+{
+	TRACE("ui_step = %d", ctx->ui_step);
+	ASSERT_UI_SCRIPT_TYPE_SANITY();
+
+	ui_callback_fn_t* this_fn = deriveScriptHash_display_ui_runStep;
+	UI_STEP_BEGIN(ctx->ui_step, this_fn);
+
+	UI_STEP(DISPLAY_UI_STEP_COMPLEX_SCRIPT_FINISHED) {
+		// For complex script finished we only show one screen, so we directly jump
+		// here to the last respond step
+		deriveScriptHash_display_ui_complexScriptFinished(this_fn);
+		// Override the next step
+		*__ui_step_ptr = DISPLAY_UI_STEP_RESPOND;
+		break;
+	}
+
+	UI_STEP(DISPLAY_UI_STEP_POSITION) {
+		// For complex scripts we reduce the current level by 1
+		// Because they already have the level increased by 1
+		int levelOffset = ctx->ui_scriptType == UI_SCRIPT_ALL
+		                  || ctx->ui_scriptType == UI_SCRIPT_ANY
+		                  || ctx->ui_scriptType == UI_SCRIPT_N_OF_K
+		                  ? -1 : 0;
+		uint8_t level = ctx->level + levelOffset;
+		// if we are on the root level don't show the position
```

Comment:
I'd wrap this logic into a function `_shouldShowScriptPosition`.

Reply:
Has this been done? I don't see it in the current version.

Reply:
Sorry forgot to push the commit.

Reply:
Extracted to a different function, because the level is needed later.